testplan/runnable/base.py

```diff
@@ -393,106 +394,124 @@ def schedule(self, task=None, resource=None, uid=None, **options):
+            :py:class:`~testplan.runners.pools.tasks.base.Task` or ``callable``
         :param resource: Test executor resource.
         :type resource: :py:class:`~testplan.runners.base.Executor`
-        :param uid: Optional test uid.
-        :type uid: ``str``
         :return: Assigned uid for test.
-        :rtype: ``str``
+        :rtype: ``str`` or ```NoneType``
         """
-        uid = uid or getattr(runnable, "uid", strings.uuid4)()
-        if uid in self._tests:
-            self.logger.error(
-                "Skip adding {} with uid {}.. already added.".format(
-                    runnable, uid
-                )
-            )
-            return uid
-
-        if isinstance(runnable, Entity):
-            runnable.cfg.parent = self.cfg
-            runnable.parent = self
-        elif isinstance(runnable, Task):
-            pass
-        elif callable(runnable):
-            runnable.parent_cfg = self.cfg
-            runnable.parent = self
-
-        # Check if test should not be added only when a filter is used.
-        if (
-            type(self.cfg.test_filter) is not filtering.Filter
-            or self.cfg.test_lister is not None
-        ):
-            if not self.should_be_added(runnable):
-                return None
-
         local_runner = self.resources.first()
-        if resource is None:
-            resource = local_runner
+        resource = resource or local_runner
 
         if resource not in self.resources:
             raise RuntimeError(
                 'Resource "{}" does not exist.'.format(resource)
             )
 
-        # When running interactively, add all tasks to the local runner even
-        # if they were scheduled into a pool. It greatly simplifies the
-        # interactive runner if it only has to deal with the local runner.
+        # Get the real test entity and verify if it should be added
+        runnable = self._verify_test_target(target)
+        if not runnable:
+            return None
+
+        uid = runnable.uid()
+        part = (
+            getattr(runnable.cfg, "part", None)
+            if hasattr(runnable, "cfg")
+            else None
+        )
+
+        # When running interactively, add all real test entities into the local
+        # runner even if they were scheduled into a pool. It greatly simplifies
+        # the interactive runner if it only has to deal with the local runner.
         if self.cfg.interactive_port is not None:
-            if isinstance(runnable, Task):
-                runnable = runnable.materialize()
-                runnable.cfg.parent = self.cfg
-                runnable.parent = self
-            self.resources[local_runner].add(runnable, runnable.uid() or uid)
-        else:
-            self.resources[resource].add(runnable, uid)
+            if part:
+                # Modify Multitest-part instance to avoid duplicate uid
+                uid = "{} - part({}/{})".format(uid, part[0], part[1])
```

Comment:
Also, shall we override multitest uid to use name+part{}/{}?

Each part of multitest shall have a different UID, and they could be merged by their name.

Reply:
updated